### PR TITLE
Support git unicode branches

### PR DESCRIPTION
**Pull request by humitos**
Use `gitpython` to get repository branches instead of our custom/hacky csv parser.

With this change, we loose the output from `git branch -r` command shown to the user under the build output.

These changes need more QA:

- repo with Unicode branch imported properly in RTD running Py2 _and_ Py3
- docs for that specific branch are serving without problem (may be affected by #1410)

NOTE: this PR also changes how the branches are parsed and I'm not 100% sure that removing the `origin/` from its name is enough and exactly the same that we were doing before

Ref: #2997 #4052
Fixes: #3060

### CODE REVIEW

**Comment by humitos** [comment]
I'm not being able to run the tests with unicode chars on branch/tag names in my local instance (`test_git_tags` also fails). I know that @stsewd worked on this issue and fixed in Travis.

What do I need to do to be able to run these unicode tests in my local instance? (I want to be able to run them all with Py2 and Py3 using tox but also make them work when just running `./manage.py runserver` so I can QA on it)

**Comment by stsewd** [comment]
I wasn't able to make this work with `manage.py runserver` (python2) I think it is because of some environment variables missing (or some additional)

**Comment by humitos** [comment]
I think we should implement something like #4442 here also as a workaround for now, since we are having different problems in production because of this issues of Unicode branches :/

**Comment by stsewd** [comment]
But we aren't close to run rtd with py3 in production?

**Comment by ericholscher** [comment]
We're now running py3 in prod, so what is the status of this PR?

**Comment by humitos** [comment]
@ericholscher Unicode branches/tags are supported in production with the current code.

Although, this branch also migrates the code to parse the output of `git branch -r` to `gitpython` which should be better than our hacky CSV parser code.

So, it's still valid for a review if we want to go in that direction.

**Comment by ericholscher** [comment]
I think it makes sense to reduce our hacky parsing when possible, especially if a library we're already using does it for us, and is almost certainly better tested and supported.

**Comment by humitos** [comment]
@stsewd I think that `test_fetch_clean_tags_and_branches` is not accurate.

I understand that the idea is to check the differences between the `upstream_repo` and the `local_repo` but,

1. `local_repo` is not a git repository, it's just a directory
1. the `repo.working_dir` points to `/tmp/tmpah5m49lq/local` but the `repo.tags` uses `/tmp/tmppsecoe9j/sample_repo` (`upstream_repo`)
    * if `/tmp/tmpah5m49lq/local` would be used, it would fail because it's not a real repo
    * if that path is not a real repo, the test doesn't make too much sense: where are these branches/tags coming from?

I don't really understand how it was working before my changes in this PR, but look at the following code:

```python
# This is the original code from master
# with a pdb at test_backend.py:L149
(Pdb) repo
<readthedocs.vcs_support.backends.git.Backend object at 0x7f4483b57978>
(Pdb) repo.working_dir
'/tmp/tmpjlndlpwd/local'
(Pdb) repo.tags
[<VCSVersion: /tmp/tmpi5ht05y4/sample_repo:v01, <VCSVersion: /tmp/tmpi5ht05y4/sample_repo:v02]
(Pdb) repo.branches
[<VCSVersion: /tmp/tmpi5ht05y4/sample_repo:invalidsubmodule, <VCSVersion: /tmp/tmpi5ht05y4/sample_repo:master, <VCSVersion: /tmp/tmpi5ht05y4/sample_repo:newbranch, <VCSVersion: /tmp/tmpi5ht05y4/sample_repo:relativesubmodule, <VCSVersion: /tmp/tmpi5ht05y4/sample_repo:submodule]
(Pdb) self.project.repo
'/tmp/tmpi5ht05y4/sample_repo'
(Pdb) import git
(Pdb) git.Repo(self.project.repo).tags
[<git.TagReference "refs/tags/v01">]
(Pdb) git.Repo(self.project.repo).branches
[<git.Head "refs/heads/invalidsubmodule">, <git.Head "refs/heads/master">, <git.Head "refs/heads/relativesubmodule">, <git.Head "refs/heads/submodule">]
(Pdb) local_repo
'/tmp/tmpjlndlpwd/local'
(Pdb) 
```

It doesn't make sense to me. The repo is not using the `local_repo` path at all to get the branches/tags. It's always from `upstream_repo` and there is somehow a cache that is killing us.

**Comment by stsewd** [comment]
I think you are missing the clone step?

https://github.com/rtfd/readthedocs.org/blob/c37b00995c1bbc5ee51d3552ef176546373bb912/readthedocs/rtd_tests/tests/test_backend.py#L150-L150

there we clone the local repo, I'll pdb to be sure.

**Comment by stsewd** [comment]
I pdb and found that local path is getting a repo

```bash
/tmp/tmp0d8n7ki5/local on  master ⌚ 16:49:23
$ g status 
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean

```

Also the remote_repo

```bash
/tmp/tmpmfuupwhq/sample_repo on  master! ⌚ 16:50:44
$ g status
On branch master
Untracked files:
  (use "git add <file>..." to include in what will be committed)

	invalidsubmodule/
	relativesubmodule/

nothing added to commit but untracked files present (use "git add" to track)

```

repo.tags uses local repo (which is the same frmo working_dir)

https://github.com/rtfd/readthedocs.org/blob/c37b00995c1bbc5ee51d3552ef176546373bb912/readthedocs/vcs_support/backends/git.py#L159-L163

**Comment by stsewd** [comment]
Not sure what is going on with gitpython and branches, I'll investigate

File: readthedocs/vcs_support/backends/git.py [hunk 1]
```diff
@@ -174,51 +172,23 @@ def tags(self):
...
-                    verbose_name = branch.replace('origin/', '')
-                    if verbose_name in ['HEAD']:
-                        continue
-                    clean_branches.append(
-                        VCSVersion(self, branch, verbose_name))
-                else:
-                    clean_branches.append(VCSVersion(self, branch, branch))
-        return clean_branches
+        # ``repo.branches`` returns local branches and
+        branches = repo.branches
```
[the remark this answers is not in the document]
Comment (humitos) [reply readthedocs/vcs_support/backends/git.py]: Our tests were expecting the local branches also, like: `master` or `develop`.

It seems that `branch -r` doesn't work exactly as `repo.origin.refs` from gitpython.

I think we are OK by using both with gitpython.

**Comment by humitos** [comment]
This PR is ready to merge after review.

**Pull request merged**